### PR TITLE
[SPARK-6994] Allow to fetch field values by name in sql.Row

### PR DESCRIPTION
**Pull request by vidma**
It looked weird that up to now there was no way in Spark's Scala API to access fields of `DataFrame/sql.Row` by name, only by their index.

This tries to solve this issue.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/RowImplicits.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+object RowImplicits {
+
+  implicit class RowWithNamedFields[K, V](sqlRow: Row) {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/RowImplicits.scala]: why don't we just add this to Row itself?

Comment (vidma) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/RowImplicits.scala]: :+1: if so, I guess I should rename `getValue[T]` into an overloaded `getAs[T](fieldName: String)` to be consistent with an existing method `getAs`? (I remember overloading didn't work with implicits...)

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/RowImplicits.scala]: yup sounds good

Comment (vidma) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/RowImplicits.scala]: For the new method to be convenient to use it should go into `Row`. 
However, the schema is not always defined for a `Row`, so `Row.getAs[T](fieldName: String)` with throw an `UnsupportedOperationException`.
`GenericRowWithSchema` will implement the case when schema exists.

**Comment by vidma** [comment]
moved the changes to the `Row` and `RowWithSchema` classes, and also added a `fieldIndex` to the schema (`StructType`).

also got rid of dependency on HiveTest.

**Comment by vidma** [comment]
it's slightly weird that there exists a `RowSuite.scala` in `sql-core` and none in `sql-catalyst`.
Anyway, the one in `sql-core` looks like integration test, so I added a little test there too, to make sure it works on Rows created via DataFrames.

**Comment by marmbrus** [comment]
ok to test

**Comment by marmbrus** [comment]
Thanks, merged to master.

**Pull request closed without merging**